### PR TITLE
feat: new goTestJunit step

### PR DESCRIPTION
**Pull request by kuisathaverat**
## What does this PR do?

A new step to run Go Unit tests and generate a JUnit report.

## Why is it important?

It simplifies/unifies the Go test executions across projects.

### CODE REVIEW

**Review by cachedout — APPROVED** [review]

File: vars/goTestJunit.groovy [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+ goTestJunit(options: '-v ./...', output: 'build/junit-report.xml')
+
+*/
+def call(Map args = [:]) {
+  def options = args.containsKey('options') ? args.options : ''
+  def output = args.containsKey('output') ? args.output : 'junit-report.xml'
+  log(level: 'INFO', text: 'Running Go test an generating JUnit output' + text)
+
+  withGoEnv(pkgs: ["github.com/jstemmer/go-junit-report"]){
```
Comment (mdelapenya) [inline vars/goTestJunit.groovy:29]: Should the package be `gotest.tools/gotestsum`?
```suggestion
  withGoEnv(pkgs: ["gotest.tools/gotestsum"]){
```

File: vars/goTestJunit.groovy [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    cmd(label: 'Running Go tests' , script: "gotestsum --junitfile ${output} -- -v ${options}")
```
Comment (mdelapenya) [inline vars/goTestJunit.groovy:30]: Also, to improve the output in Jenkins logs, if we set `--format testname` instead, and we set an env var `GOTESTSUM_JUNITFILE=${output}` then we have both things at the same time: good log and jUnit file.

Comment (kuisathaverat) [reply vars/goTestJunit.groovy]: I prefer to use only options or only env vars

**Review by mdelapenya — APPROVED** [review]

File: vars/withGoEnv.groovy [hunk 3]
```diff
@@ -44,3 +45,9 @@ def call(Map args = [:], Closure body) {
     }
   }
 }
+
+def checkGoPath(){
+  if(fileExists(file: "${env.GOPATH}/go.mod")){
```
Comment (v1v) [inline vars/withGoEnv.groovy:50]: is `GOPATH` defined earlier? If so, where?

File: vars/goTestJunit.groovy [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+ goTestJUnit(options: '-v ./...', output: 'build/junit-report.xml')
+
+*/
+def call(Map args = [:]) {
+  def options = args.containsKey('options') ? args.options : ''
+  def output = args.containsKey('output') ? args.output : 'junit-report.xml'
+  log(level: 'INFO', text: 'Running Go test an generating JUnit output')
+
+  withGoEnv(pkgs: ["gotest.tools/gotestsum"]){
```
Comment (v1v) [inline vars/goTestJunit.groovy:29]: I wonder if the `go-version` should be also somehow customised or enabled to be customised?

File: vars/withGoEnv.groovy (hunk 3 above)
Comment (kuisathaverat) [reply vars/withGoEnv.groovy:50]: it is defined by `withGoEnv`

File: vars/goTestJunit.groovy (hunk 4 above)
Comment (kuisathaverat) [reply vars/goTestJunit.groovy:29]: it is you can use the GO_VERSION environment variable.

Comment (kuisathaverat) [reply vars/goTestJunit.groovy]: in any case, I've added it.

File: vars/withGoEnv.groovy (hunk 3 above)
Comment (v1v) [reply vars/withGoEnv.groovy:50]: right, but it does not exist yet when the checkGoPath is called

**Review by v1v — CHANGES_REQUESTED** [review]
See my comment regarding the `checkGoPath` method

File: vars/withGoEnv.groovy (hunk 3 above)
Comment (kuisathaverat) [reply vars/withGoEnv.groovy:51]: I've moved the call inside the `withGoEnvUnix` and `withGoEnvWindows`  closures

**Review by v1v — APPROVED** [review]

**Pull request merged**